third_party/move/move-model/src/builder/model_builder.rs

```diff
@@ -454,6 +451,121 @@ impl<'env> ModelBuilder<'env> {
         }
     }
 
+    /// Gets the name of the struct
+    fn get_struct_name(&self, qid: QualifiedId<StructId>) -> &QualifiedSymbol {
+        self
+            .reverse_struct_table
+            .get(&(qid.module_id, qid.id))
+            .expect("invalid Type::Struct")
+    }
+
+    /// Gets the simple display name of the struct
+    fn get_struct_display_name_simple(&self, qid: QualifiedId<StructId>) -> String {
+        self.get_struct_name(qid).display_simple(self.env).to_string()
+    }
+
+    /// Gets the display name of the struct type
+    /// Requires: the type is a struct type
+    fn get_struct_type_name(&self, ty: &Type) -> String {
+        if let Type::Struct(mid, sid, _) = ty {
+            // TODO: display the struct with type parameters
+            let qid = QualifiedId { module_id: *mid, id: *sid };
+            self.get_struct_display_name_simple(qid).to_string()
+        } else {
+            panic!("ICE: calling `get_struct_type_name` with non-struct type")
+        }
+    }
+
+    /// Generates error messages for recursive structs
+    /// `path`: `path[0]` contains `path[1]` ... `path[-1]` contains `this_struct`
+    fn gen_error_msg_for_fields_loop(
+        &self,
+        path: &Vec<Type>,
+        this_struct_id: QualifiedId<StructId>
+    ) -> Vec<String> {
+        let mut loop_notes = Vec::new();
+        for i in 0..path.len() - 1 {
+            let parent_name = self.get_struct_type_name(&path[i]);
+            let child_name = self.get_struct_type_name(&path[i + 1]);
+            loop_notes.push(format!("{} contains {}...", parent_name, child_name));
+        }
+        loop_notes.push(format!("{} contains {}, which forms a loop.", self.get_struct_type_name(path.last().unwrap()), self.get_struct_display_name_simple(this_struct_id)));
```

Comment:
Please align with other error messages. Our convention is to put programming symbols into backticks, hence: "\backquote{}\backquote contains..."

Reply:
done